File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/jdbc/JdbcRepositoryGenerator.java
package ru.tinkoff.kora.database.annotation.processor.jdbc;

import com.squareup.javapoet.*;
import reactor.core.publisher.Mono;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.FieldFactory;
import ru.tinkoff.kora.annotation.processor.common.Visitors;
import ru.tinkoff.kora.common.Tag;
import ru.tinkoff.kora.database.annotation.processor.DbUtils;
import ru.tinkoff.kora.database.annotation.processor.DbUtils.Mapper;
import ru.tinkoff.kora.database.annotation.processor.QueryWithParameters;
import ru.tinkoff.kora.database.annotation.processor.RepositoryGenerator;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameter;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameterParser;

import javax.annotation.Nullable;
import javax.annotation.processing.Filer;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.stream.IntStream;

import static ru.tinkoff.kora.annotation.processor.common.MethodUtils.isVoid;

public final class JdbcRepositoryGenerator implements RepositoryGenerator {
    private final TypeMirror repositoryInterface;
    private final Types types;
    private final Elements elements;
    private final Filer filer;

    public JdbcRepositoryGenerator(ProcessingEnvironment processingEnv) {
        var repository = processingEnv.getElementUtils().getTypeElement(JdbcTypes.JDBC_REPOSITORY.canonicalName());
        if (repository == null) {
            this.repositoryInterface = null;
        } else {
            this.repositoryInterface = repository.asType();
        }
        this.types = processingEnv.getTypeUtils();
        this.elements = processingEnv.getElementUtils();
        this.filer = processingEnv.getFiler();
    }

    @Override
    public TypeSpec generate(TypeElement repositoryElement, TypeSpec.Builder type, MethodSpec.Builder constructor) {
        var repositoryType = (DeclaredType) repositoryElement.asType();
        var queryMethods = DbUtils.findQueryMethods(this.types, this.elements, repositoryElement);
        this.enrichWithExecutor(repositoryElement, type, constructor, queryMethods);
        var resultMappers = new FieldFactory(this.types, type, constructor, "_result_mapper_");
        var parameterMappers = new FieldFactory(this.types, type, constructor, "_parameter_mapper_");
        for (var method : queryMethods) {
            var methodType = (ExecutableType) this.types.asMemberOf(repositoryType, method);
            var parameters = QueryParameterParser.parse(this.types, JdbcTypes.CONNECTION, method, methodType);
            var queryAnnotation = CommonUtils.findDirectAnnotation(method, DbUtils.QUERY_ANNOTATION);
            var queryString = CommonUtils.parseAnnotationValueWithoutDefault(queryAnnotation, "value").toString();
            var query = QueryWithParameters.parse(filer, queryString, parameters);
            var resultMapper = this.parseResultMapper(method, methodType, parameters)
                .map(rm -> DbUtils.addMapper(resultMappers, rm))
                .orElse(null);
            DbUtils.addMappers(parameterMappers, DbUtils.parseParameterMappers(
                parameters,
                query,
                tn -> JdbcNativeTypes.findNativeType(tn) != null,
                JdbcTypes.PARAMETER_COLUMN_MAPPER
            ));
            var methodSpec = this.generate(method, methodType, query, parameters, resultMapper, parameterMappers);
            type.addMethod(methodSpec);
        }
        return type.addMethod(constructor.build()).build();
    }

    private Optional<Mapper> parseResultMapper(ExecutableElement method, ExecutableType methodType, List<QueryParameter> parameters) {
        var returnType = methodType.getReturnType();
        if (CommonUtils.isMono(returnType)) {
            returnType = Visitors.visitDeclaredType(returnType, dt -> dt.getTypeArguments().get(0));
        }
        if (isVoid(returnType)) {
            return Optional.empty();
        }
        var batchParam = parameters.stream().filter(QueryParameter.BatchParameter.class::isInstance).findFirst().orElse(null);
        if (batchParam != null) {
            // either void or update count, no way to parse results from db with jdbc api
            return Optional.empty();
        }
        if (returnType.toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
            return Optional.empty();
        }
        var mappings = CommonUtils.parseMapping(method);
        var mapperType = ParameterizedTypeName.get(
            JdbcTypes.RESULT_SET_MAPPER, TypeName.get(returnType).box()
        );
        var resultSetMapper = mappings.getMapping(JdbcTypes.RESULT_SET_MAPPER);
        if (resultSetMapper != null) {
            return Optional.of(new Mapper(resultSetMapper.mapperClass(), mapperType, mappings.mapperTags()));
        }
        var rowMapper = mappings.getMapping(JdbcTypes.ROW_MAPPER);
        if (rowMapper != null) {
            if (CommonUtils.isList(returnType)) {
                return Optional.of(new Mapper(rowMapper.mapperClass(), mapperType, mappings.mapperTags(), c -> CodeBlock.of("$T.listResultSetMapper($L)", JdbcTypes.RESULT_SET_MAPPER, c)));
            } else if (CommonUtils.isOptional(returnType)) {
                return Optional.of(new Mapper(rowMapper.mapperClass(), mapperType, mappings.mapperTags(), c -> CodeBlock.of("$T.optionalResultSetMapper($L)", JdbcTypes.RESULT_SET_MAPPER, c)));
            } else {
                return Optional.of(new Mapper(rowMapper.mapperClass(), mapperType, mappings.mapperTags(), c -> CodeBlock.of("$T.singleResultSetMapper($L)", JdbcTypes.RESULT_SET_MAPPER, c)));
            }
        }
        return Optional.of(new Mapper(mapperType, mappings.mapperTags()));
    }

    @Override
    @Nullable
    public TypeMirror repositoryInterface() {
        return this.repositoryInterface;
    }

    public MethodSpec generate(ExecutableElement method, ExecutableType methodType, QueryWithParameters query, List<QueryParameter> parameters, @Nullable String resultMapperName, FieldFactory parameterMappers) {
        var batchParam = parameters.stream().filter(QueryParameter.BatchParameter.class::isInstance).findFirst().orElse(null);
        var sql = query.rawQuery();
        for (var parameter : query.parameters().stream().sorted(Comparator.<QueryWithParameters.QueryParameter>comparingInt(s -> s.sqlParameterName().length()).reversed()).toList()) {
            sql = sql.replace(":" + parameter.sqlParameterName(), "?");
        }

        var b = DbUtils.queryMethodBuilder(method, methodType);
        final boolean isMono = CommonUtils.isMono(methodType.getReturnType());
        if (isMono) {
            b.addCode("return $T.fromCompletionStage(() -> $T.supplyAsync(() -> {$>\n", Mono.class, CompletableFuture.class);
        }
        var connection = parameters.stream().filter(QueryParameter.ConnectionParameter.class::isInstance).findFirst()
            .map(p -> CodeBlock.of("$L", p.variable()))
            .orElse(CodeBlock.of("this._connectionFactory.currentConnection()"));
        b.addCode("""
            var _conToUse = $L;
            $T _conToClose;
            if (_conToUse == null) {
                _conToUse = this._connectionFactory.newConnection();
                _conToClose = _conToUse;
            } else {
                _conToClose = null;
            }
            var _query = new $T(
              $S,
              $S
            );
            var _telemetry = this._connectionFactory.telemetry().createContext(ru.tinkoff.kora.common.Context.current(), _query);
            try (_conToClose; var _stmt = _conToUse.prepareStatement(_query.sql())) {$>
            """, connection, JdbcTypes.CONNECTION, DbUtils.QUERY_CONTEXT, query.rawQuery(), sql);
        b.addCode(StatementSetterGenerator.generate(method, query, parameters, batchParam, parameterMappers));
        if (isVoid(method) || isMono && isVoid(((DeclaredType) methodType.getReturnType()).getTypeArguments().get(0))) {
            if (batchParam != null) {
                b.addStatement("var _batchResult = _stmt.executeBatch()");
            } else {
                b
                    .addStatement("_stmt.execute();")
                    .addStatement("var updateCount = _stmt.getUpdateCount()");
            }
            b.addStatement("_telemetry.close(null)");
            if (isMono) {
                b.addStatement("return null");
            }
        } else if (batchParam != null) {
            b.addStatement("var _batchResult = _stmt.executeBatch()");
            b.addStatement("_telemetry.close(null)");
            if (methodType.getReturnType().toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
                b.addStatement("return new $T($T.of(_batchResult).sum())", DbUtils.UPDATE_COUNT, IntStream.class);
            } else {
                b.addStatement("return _batchResult");
            }
        } else if (methodType.getReturnType().toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
            b
                .addStatement("var _updateCount = _stmt.executeLargeUpdate()")
                .addStatement("_telemetry.close(null)")
                .addStatement("return new $T(_updateCount)", DbUtils.UPDATE_COUNT);
        } else {
            var result = CommonUtils.isNullable(method) || method.getReturnType().getKind().isPrimitive() || isMono
                ? CodeBlock.of("_result")
                : CodeBlock.of("$T.requireNonNull(_result)", Objects.class);
            b.addCode("try (var _rs = _stmt.executeQuery()) {$>\n")
                .addCode("var _result = $L.apply(_rs);\n", resultMapperName)
                .addCode("_telemetry.close(null);\n")
                .addCode("return $L;", result)
                .addCode("$<\n}\n");
        }
        b.addCode("$<\n} catch (java.sql.SQLException e) {\n")
            .addCode("  _telemetry.close(e);\n")
            .addCode("  throw new ru.tinkoff.kora.database.jdbc.RuntimeSqlException(e);\n")
            .addCode("}  catch (Exception e) {\n")
            .addCode("  _telemetry.close(e);\n")
            .addCode("  throw e;\n")
            .addCode("}\n");

        if (isMono) {
            b.addCode("$<\n}));\n");
        }
        return b.build();
    }

    public void enrichWithExecutor(TypeElement repositoryElement, TypeSpec.Builder builder, MethodSpec.Builder constructorBuilder, List<ExecutableElement> queryMethods) {
        builder.addField(JdbcTypes.CONNECTION_FACTORY, "_connectionFactory", Modifier.PRIVATE, Modifier.FINAL);
        builder.addSuperinterface(JdbcTypes.JDBC_REPOSITORY);
        builder.addMethod(MethodSpec.methodBuilder("getJdbcConnectionFactory")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addAnnotation(Override.class)
            .returns(JdbcTypes.CONNECTION_FACTORY)
            .addCode("return this._connectionFactory;")
            .build());

        var executorTag = DbUtils.getTag(repositoryElement);
        if (executorTag != null) {
            constructorBuilder.addParameter(ParameterSpec.builder(JdbcTypes.CONNECTION_FACTORY, "_connectionFactory").addAnnotation(AnnotationSpec.builder(Tag.class).addMember("value", executorTag).build()).build());
        } else {
            constructorBuilder.addParameter(JdbcTypes.CONNECTION_FACTORY, "_connectionFactory");
        }
        constructorBuilder.addStatement("this._connectionFactory = _connectionFactory");

        var needThreadPool = queryMethods.stream().anyMatch(e -> CommonUtils.isMono(e.getReturnType()));
        if (needThreadPool && executorTag != null) {
            builder.addField(TypeName.get(Executor.class), "_executor", Modifier.PRIVATE, Modifier.FINAL);
            constructorBuilder.addStatement("this._executor = _executor");
            constructorBuilder.addParameter(ParameterSpec.builder(TypeName.get(Executor.class), "_executor").addAnnotation(AnnotationSpec.builder(Tag.class).addMember("value", executorTag).build()).build());
        }
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/jdbc/JdbcRepositoryGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.jdbc

import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunction
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toTypeName
import reactor.core.publisher.Mono
import ru.tinkoff.kora.database.symbol.processor.DbUtils
import ru.tinkoff.kora.database.symbol.processor.DbUtils.addMapper
import ru.tinkoff.kora.database.symbol.processor.DbUtils.awaitSingle
import ru.tinkoff.kora.database.symbol.processor.DbUtils.awaitSingleOrNull
import ru.tinkoff.kora.database.symbol.processor.DbUtils.findQueryMethods
import ru.tinkoff.kora.database.symbol.processor.DbUtils.parseExecutorTag
import ru.tinkoff.kora.database.symbol.processor.DbUtils.queryMethodBuilder
import ru.tinkoff.kora.database.symbol.processor.DbUtils.resultMapperName
import ru.tinkoff.kora.database.symbol.processor.DbUtils.updateCount
import ru.tinkoff.kora.database.symbol.processor.Mapper
import ru.tinkoff.kora.database.symbol.processor.QueryWithParameters
import ru.tinkoff.kora.database.symbol.processor.RepositoryGenerator
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameter
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameterParser
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.CommonClassNames.isList
import ru.tinkoff.kora.ksp.common.FieldFactory
import ru.tinkoff.kora.ksp.common.FunctionUtils.isSuspend
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.parseMappingData
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor

class JdbcRepositoryGenerator(private val resolver: Resolver) : RepositoryGenerator {
    private val repositoryInterface = resolver.getClassDeclarationByName(resolver.getKSNameFromString(JdbcTypes.jdbcRepository.canonicalName))?.asStarProjectedType()
    override fun repositoryInterface() = repositoryInterface

    override fun generate(repositoryType: KSClassDeclaration, typeBuilder: TypeSpec.Builder, constructorBuilder: FunSpec.Builder): TypeSpec {
        val queryMethods = repositoryType.findQueryMethods()
        enrichWithExecutor(repositoryType, typeBuilder, constructorBuilder, queryMethods)
        val repositoryResolvedType = repositoryType.asStarProjectedType()
        val resultMappers = FieldFactory(typeBuilder, constructorBuilder, "_result_mapper_")
        val parameterMappers = FieldFactory(typeBuilder, constructorBuilder, "_parameter_mapper_")
        for (method in queryMethods) {
            val methodType = method.asMemberOf(repositoryResolvedType)
            val parameters = QueryParameterParser.parse(JdbcTypes.connection, JdbcTypes.jdbcParameterColumnMapper, method, methodType)
            val queryAnnotation = method.findAnnotation(DbUtils.queryAnnotation)!!
            val queryString = queryAnnotation.findValue<String>("value")!!
            val query = QueryWithParameters.parse(queryString, parameters)
            val resultMapper = this.parseResultMapper(method, parameters, methodType)?.let { resultMappers.addMapper(it) }
            DbUtils.parseParameterMappers(method, parameters, query, JdbcTypes.jdbcParameterColumnMapper) { JdbcNativeTypes.findNativeType(it.toTypeName()) != null }
                .forEach { parameterMappers.addMapper(it) }
            val methodSpec = this.generate(method, methodType, query, parameters, resultMapper, parameterMappers)
            typeBuilder.addFunction(methodSpec)
        }
        return typeBuilder.primaryConstructor(constructorBuilder.build()).build()
    }

    private fun generate(method: KSFunctionDeclaration, methodType: KSFunction, query: QueryWithParameters, parameters: List<QueryParameter>, resultMapperName: String?, parameterMappers: FieldFactory): FunSpec {
        val batchParam = parameters.firstOrNull { it is QueryParameter.BatchParameter }
        var sql = query.rawQuery
        for (parameter in query.parameters.sortedByDescending { it.sqlParameterName.length }) {
            sql = sql.replace(":${parameter.sqlParameterName}", "?")
        }
        val b = method.queryMethodBuilder(resolver)
        if (method.isSuspend()) {
            b.beginControlFlow("return %T.fromCompletionStage({", Mono::class)
            b.beginControlFlow("%T.supplyAsync({", CompletableFuture::class)
        }
        val returnTypeName = methodType.returnType?.toTypeName()

        val connection = parameters.firstOrNull { it is QueryParameter.ConnectionParameter }
            ?.let { CodeBlock.of("%L", it.variable) } ?: CodeBlock.of("this._jdbcConnectionFactory.currentConnection()")
        b.addStatement("var _conToUse = %L", connection)
        b.addStatement("val _conToClose: %T?", JdbcTypes.connection)
        b.controlFlow("if (_conToUse == null)") {
            addStatement("_conToUse = this._jdbcConnectionFactory.newConnection()")
            addStatement("_conToClose = _conToUse")
            nextControlFlow("else")
            addStatement("_conToClose = null")
        }
        b.addStatement("val _query = %T(%S, %S)", DbUtils.queryContext, query.rawQuery, sql)
        b.addStatement("val _telemetry = this._jdbcConnectionFactory.telemetry().createContext(ru.tinkoff.kora.common.Context.current(), _query)")
        b.controlFlow("try") {
            controlFlow("_conToClose.use") {
                controlFlow("_conToUse!!.prepareStatement(_query.sql()).use { _stmt ->") {
                    StatementSetterGenerator.generate(b, method, query, parameters, batchParam, parameterMappers)
                    if (methodType.returnType!! == resolver.builtIns.unitType) {
                        if (batchParam != null) {
                            addStatement("_stmt.executeBatch()")
                        } else {
                            addStatement("_stmt.execute()")
                            addStatement("_stmt.getUpdateCount()")
                        }
                        addStatement("_telemetry.close(null)")
                    } else if (returnTypeName == updateCount) {
                        if (batchParam != null) {
                            addStatement("val _updateCount = _stmt.executeLargeBatch().sum()")
                        } else {
                            addStatement("val _updateCount = _stmt.executeLargeUpdate()")
                        }
                        addStatement("_telemetry.close(null)")
                        addCode("return")
                        if (method.isSuspend()) {
                            addCode("@supplyAsync")
                        }
                        addCode(" %T(_updateCount)\n", updateCount)
                    } else {
                        controlFlow("_stmt.executeQuery().use { _rs ->") {
                            addStatement("val _result = %N.apply(_rs)", resultMapperName!!)
                            addStatement("_telemetry.close(null)")
                            addCode("return")
                            if (method.isSuspend()) {
                                addCode("@supplyAsync")
                            }
                            addCode(" _result")
                            if (!methodType.returnType!!.isMarkedNullable) {
                                addCode("!!")
                            }
                            addCode("\n")
                        }
                    }
                }
            }
            nextControlFlow("catch (_e: java.sql.SQLException)")
            addStatement("_telemetry.close(_e)")
            addStatement("throw ru.tinkoff.kora.database.jdbc.RuntimeSqlException(_e)")
            nextControlFlow("catch (_e: Exception)")
            addStatement("_telemetry.close(_e)")
            addStatement("throw _e")
        }
        if (method.isSuspend()) {
            b.endControlFlow()
            b.addCode(", this._executor)")
            b.endControlFlow()
            b.addCode(")")
            if (methodType.returnType!!.isMarkedNullable) {
                b.addCode(".%M()", awaitSingleOrNull)
            } else {
                b.addCode(".%M()", awaitSingle)
            }
        }
        return b.build()
    }

    private fun parseResultMapper(method: KSFunctionDeclaration, parameters: List<QueryParameter>, methodType: KSFunction): Mapper? {
        for (parameter in parameters) {
            if (parameter is QueryParameter.BatchParameter) {
                return null
            }
        }
        val returnType = methodType.returnType!!
        val mapperName = method.resultMapperName()
        val mappings = method.parseMappingData()
        val resultSetMapper = mappings.getMapping(JdbcTypes.jdbcResultSetMapper)
        val rowMapper = mappings.getMapping(JdbcTypes.jdbcRowMapper)
        val returnTypeName = returnType.toTypeName().copy(false)
        val mapperType = JdbcTypes.jdbcResultSetMapper.parameterizedBy(returnTypeName)
        if (resultSetMapper != null) {
            return Mapper(resultSetMapper, mapperType, mapperName)
        }
        if (rowMapper != null) {
            if (returnType.isList()) {
                return Mapper(rowMapper, mapperType, mapperName) {
                    CodeBlock.of("%T.listResultSetMapper(%L)", JdbcTypes.jdbcResultSetMapper, it)
                }
            } else {
                return Mapper(rowMapper, mapperType, mapperName) {
                    CodeBlock.of("%T.singleResultSetMapper(%L)", JdbcTypes.jdbcResultSetMapper, it)
                }
            }
        }
        if (returnType == resolver.builtIns.unitType) {
            return null
        }
        if (returnTypeName == updateCount) {
            return null
        }
        return Mapper(mapperType, mapperName)
    }

    private fun enrichWithExecutor(repositoryElement: KSClassDeclaration, builder: TypeSpec.Builder, constructorBuilder: FunSpec.Builder, queryMethods: Sequence<KSFunctionDeclaration>) {
        builder.addProperty("_jdbcConnectionFactory", JdbcTypes.connectionFactory, KModifier.PRIVATE)
        builder.addSuperinterface(JdbcTypes.jdbcRepository)
        builder.addFunction(
            FunSpec.builder("getJdbcConnectionFactory")
                .addModifiers(KModifier.OVERRIDE)
                .returns(JdbcTypes.connectionFactory)
                .addStatement("return this._jdbcConnectionFactory")
                .build()
        )
        val executorTag = repositoryElement.parseExecutorTag()
        if (executorTag != null) {
            constructorBuilder.addParameter(
                ParameterSpec.builder("_jdbcConnectionFactory", JdbcTypes.connectionFactory).addAnnotation(
                    AnnotationSpec.builder(CommonClassNames.tag).addMember("value = %L", executorTag).build()
                ).build()
            )
        } else {
            constructorBuilder.addParameter("_jdbcConnectionFactory", JdbcTypes.connectionFactory)
        }
        constructorBuilder.addStatement("this._jdbcConnectionFactory = _jdbcConnectionFactory")

        if (queryMethods.any { it.isSuspend() }) {
            val executor = Executor::class.asClassName()
            builder.addProperty("_executor", executor, KModifier.PRIVATE, KModifier.FINAL)
            constructorBuilder.addStatement("this._executor = _executor")
            if (executorTag != null) {
                constructorBuilder.addParameter(
                    ParameterSpec.builder("_executor", executor).addAnnotation(
                        AnnotationSpec.builder(CommonClassNames.tag).addMember("value = %L", executorTag).build()
                    ).build()
                )
            } else {
                constructorBuilder.addParameter(ParameterSpec.builder("_executor", executor).build())
            }
        }
    }
}
